File: hw/vargas/u3/hw29Singleton/javaScript/ec/edu/espe/singleton/utils/AppointmentNotificationService.js
import nodemailer from 'nodemailer';
import { GmailConfiguration } from './GmailConfiguration.js';

export class AppointmentNotificationService {
    constructor() {
        this.config = GmailConfiguration.getInstance();
    }

    async init() {
        await this.config.loadConfigurationFromDatabase();
        
        this.transporter = nodemailer.createTransport({
            host: this.config.smtpHost,
            port: this.config.smtpPort,
            secure: true, 
            auth: {
                user: this.config.senderEmail,
                pass: this.config.appPassword,
            },
        });
    }

    async sendReservationConfirmation(recipientEmail, patientName, appointmentDetails) {
        if (!this.config.senderEmail || !this.config.appPassword) {
            console.error("Error: Credenciales no configuradas.");
            return;
        }

        const mailOptions = {
            from: this.config.senderEmail,
            to: recipientEmail,
            subject: 'Confirmación de Cita Médica - Clínica Toamedical',
            text: this.#buildEmailBody(patientName, appointmentDetails)
        };

        try {
            await this.transporter.sendMail(mailOptions);
            console.log(`Notificación enviada correctamente a: ${recipientEmail}`);
        } catch (error) {
            console.error(`Error al enviar el correo: ${error.message}`);
        }
    }

    #buildEmailBody(name, details) {
        return `Estimado/a ${name},

Su cita ha sido reservada exitosamente.
Fecha de la cita agendada: ${details}

Por favor, llegue 10 minutos antes.
Atentamente,
Tu clínica de confianza Toamedical.`;
    }
}

File: hw/vargas/u3/hw29Singleton/python/ec/edu/espe/singleton/utils/appointment_notification_service.py
import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from .gmail_configuration import GmailConfiguration

class AppointmentNotificationService:

    def __init__(self):
        config = GmailConfiguration.get_instance()
        self.sender_email = config.sender_email
        self.app_password = config.app_password
        self.smtp_host = config.smtp_host
        self.smtp_port = config.smtp_port

    def send_reservation_confirmation(self, recipient_email: str, patient_name: str, appointment_details: str):
        if not self.sender_email or not self.app_password:
            print("Error: No hay credenciales de correo configuradas.")
            return

        try:
            msg = MIMEMultipart()
            msg['From'] = self.sender_email
            msg['To'] = recipient_email
            msg['Subject'] = "Confirmación de Cita Médica - Clínica Toamedical"

            body = self._build_email_body(patient_name, appointment_details)
            msg.attach(MIMEText(body, 'plain'))

            # Conexión SSL
            server = smtplib.SMTP_SSL(self.smtp_host, self.smtp_port)
            server.login(self.sender_email, self.app_password)
            server.send_message(msg)
            server.quit()

            print(f"Notificación enviada correctamente a: {recipient_email}")

        except Exception as e:
            print(f"Error al enviar el correo: {e}")

    def _build_email_body(self, name: str, details: str) -> str:
        return (
            f"Estimado/a {name},\n\n"
            "Su cita ha sido reservada exitosamente.\n"
            f"Fecha de la cita agendada: {details}\n\n"
            "Por favor, llegue 10 minutos antes.\n"
            "Atentamente,\n"
            "Tu clínica de confianza Toamedical."
        )